README.txt
== UnQlocked by Garrett a.k.a. garbear ==

UnQlocked is a script to display the current time using common phrases instead
of numbers. Originally an unabashed ripoff of Biegert & Funk's awesome product,
this script has been extended to support a much larger set of languages and
configurations.

An up-to-date copy of the code can be found at the project's GitHub page:
https://github.com/garbear/unqlocked

Limited support is available on the XBMC forum. You may also find additional
clocks and themes:
http://forum.xbmc.org/showthread.php?tid=135578


== How to run ==
XBMC Frodo will feature Python screensaver support. When you upgrade your XBMC
install to a compatable version, script.unqlocked will show up in your list of
screensavers. Until then, this script can also be run from the Programs menu
like any other add-on.


== Do a trick ==
Run UnQlocked as a script and press L to cycle layouts (same dimensions only).


== Creating your own clock ==
Clocks are defined by an XML layout file in the layouts folder. In the root
<layout> node you will need three tags.
* <background>
    A comma-separated list of entities. Each entity can be multiple
    characters (such as L' or AE). Whitespace is ignore, so use multiple lines
    for clarity. The number of entities must be equal to width * height (given
    by the width and height attributes of the <background> tag).
* <strings>
    A mapping of numbers to strings. If your clock uses a number, make sure the
    strings table has an entry for it.
* <times>
    Time strings can use symbols to represent the strings found in the strings
    table. UnQlocked will identify the offset and direction of the symbol and
    propagate it forward in time, making it possible to define a large
    expression of times using a single time string.

Not everything can be accomplished with symbols, however. If your <time> string
includes no symbols, it won't be propagated forward and will only occur for
that instant. To extend such a string, you can use a "duration" attribute in
the same format as the "id" attribute. For example:
    <time id="2:00">%2h% o'clock</time>
    <time id="2:30" duration="0:10">lunch time</time>

Here, 2:25 and 2:40 would read "two o'clock", and 2:30 and 2:35 would read
"lunch time". Remember that your layout needs all these strings embedded in the
background. Also, note that the duration attribute can be used in times with
symbols as well (if the need arises).

If the "use24" attribute is set to true, UnQlocked will use a 24-hour system.

UnQlocked uses GCD to determine your clock's resolution. Seconds symbols can be
used, but no special handling is done for them, so you may need to perform lots
of copy-pasting for anything outside the "X hours Y minutes Z seconds" format.


== Creating your own theme ==
A theme is an XML file with the following tags:
* <background>
    This color (AARRGGBB) is applied fullscreen. In screensaver mode, the alpha
    channel is ignore and the background is fully opaque. Skin color names can
    also be used here (but the skin must support it, so try to stick with
    common names).
* <image>
    An optional image behind the clock. Attributes "width" and "height" can be
    specified (the image is always centered). If width and height are set to
    1280 and 720 (or omitted), the image will be fullscreen, making the
    <background> tag optional.
* <active>
* <inactive>
    The text color of active (highlighted) and inactive letters.

If you are a skin author and want a theme to be the default for your skin, post
your theme and skin ID in the forum and I'll have my people talk to your
people.

Thanks for trying UnQlocked!

changelog.txt
[B]1.0, 2012-07-07[/B]
. Initial release

unqlocked/config.py
# *  This Program is free software; you can redistribute it and/or modify
# *  it under the terms of the GNU General Public License as published by
# *  the Free Software Foundation; either version 2, or (at your option)
# *  any later version.
# *
# *  This Program is distributed in the hope that it will be useful,
# *  but WITHOUT ANY WARRANTY; without even the implied warranty of
# *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# *  GNU General Public License for more details.
# *
# *  You should have received a copy of the GNU General Public License
# *  along with XBMC; see the file COPYING.  If not, write to
# *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
# *  http://www.gnu.org/copyleft/gpl.html

from unqlocked import log, Time

import elementtree.ElementTree as ElementTree
import os
import sys
import xbmc
import xbmcaddon


class Config:
	'''
	Config is the interface between this script and most external variables,
	including addon settings and the script's XML files.
	'''
	def __init__(self):
		self.scriptId   = 'script.unqlocked'
		self.addon      = xbmcaddon.Addon(self.scriptId)
		self.addonName  = self.addon.getAddonInfo('name')
		self.cwd        = self.addon.getAddonInfo('path')
		self.author     = self.addon.getAddonInfo('author')
		self.version    = self.addon.getAddonInfo('version')
		self.profile    = xbmc.translatePath(self.addon.getAddonInfo('profile'))
		self.ssMode     = xbmc.getCondVisibility('System.ScreenSaverActive')
		#self.language   = self.addon.getLocalizedString
		self.layoutDir  = os.path.join(self.cwd, 'layouts')
		self.layoutName = self.getLayoutFile(self.layoutDir)
		self.layout     = Layout(os.path.join(self.layoutDir, self.layoutName))
		self.themeDir   = os.path.join(self.cwd, 'themes')
		self.themeName  = self.getThemeFile(self.themeDir)
		self.theme      = Theme(os.path.join(self.themeDir, self.themeName))
	
	def getLayoutFile(self, layoutDir):
		layout = self.addon.getSetting('layout')
		if layout == 'Default':
			# Choose the best layout for the language
			map = {
				'dutch':   'Dutch 10x11',
				'english': 'English 10x11',
				'french':  'French 10x11',
				'german':  'German 10x11',
				'italian': 'Italian 10x11',
				'spanish': 'Spanish 10x11'}
			language = xbmc.getLanguage()
			if language in map:
				layout = map[language]
			else:
				layout = map['english']
			self.addon.setSetting('layout', layout)
		# .xml was masked in the settings
		return layout + '.xml'
	
	def getThemeFile(self, themeDir):
		theme = self.addon.getSetting('theme')
		if theme == 'Default':
			# Choose the best theme for the current skin
			map = {
				'skin.alaska.revisited': 'Vanilla Sugar',
				'skin.confluence':       'Black Ice Tea',
				'skin.xeebo':            'Gritty Guacamole'}
			skin = xbmc.getSkinDir()
			if skin in map:
				theme = map[skin]
			else:
				theme = map['skin.confluence']
			self.addon.setSetting('theme', theme)
		# .xml was masked in the settings
		return theme + '.xml'
	
	def loadNextTheme(self):
		themes = [theme for theme in os.listdir(self.themeDir) if theme[-4:] == '.xml']
		count = len(themes)
		for i in range(count):
			if themes[i] == self.themeName:
				self.themeName = themes[(i + 1) % count]
				self.theme = Theme(os.path.join(self.themeDir, self.themeName))
				break
	
	def loadNextLayout(self):
		layouts = [layout for layout in os.listdir(self.layoutDir) if layout[-4:] == '.xml']
		count = len(layouts)
		for i in range(count):
			if layouts[i] == self.layoutName:
				self.layoutName = layouts[(i + 1) % count]
				self.layout = Layout(os.path.join(self.layoutDir, self.layoutName))
				break


class Layout:
	'''
	Load an XML layout file into several parameters:
	* self.matrix - The 2-dimensional array of background characters
	* self.height - equal to len(self.matrix)
	* self.width - equal to len(self.matrix[0])
	* self.times - dictionary of time strings
	* self.use24 - flag for a 24-hour clock
	* self.strings - dictionary for translating numbers into strings
	'''
	def __init__(self, file):
		log('Using layout: ' + os.path.basename(file))
		try:
			root = ElementTree.parse(file).getroot()
		except:
			log('Error parsing layout file!')
			raise # Let the user see the error
		try:
			background = root.find('background')
			self.height = int(background.attrib['height'])
			self.width = int(background.attrib['width'])
		except:
			log('Error: <background> tag missing/incorrect!')
			sys.exit()
		
		self.matrix = []
		entities = [char.strip().upper() for char in background.text.split(',')]
		if (self.height * self.width > len(entities)):
			log('Error: Too many characters in background (expecting %d, found %d)' % \
				(self.height * self.width, len(entities)))
			sys.exit()
		elif (self.height * self.width < len(entities)):
			log('Error: Too few characters in background (expecting %d, found %d)' % \
				(self.height * self.width, len(entities)))
			sys.exit()
		for i in range(self.height):
			self.matrix.append(entities[i * self.width : (i + 1) * self.width])
		
		timesNode = root.find('times')
		if timesNode == None:
			log('Error: <times> node not found!')
			sys.exit()
		if 'use24' in timesNode.attrib and timesNode.attrib['use24'] == 'true':
			self.use24 = True
		else:
			self.use24 = False
		
		self.times = {}
		for time in timesNode.findall('time'):
			if 'id' not in time.attrib:
				log('Warning: found <time> tag with no "id" attribute')
				continue
			key = Time(time.attrib['id'])
			# If set, store the duration with the key
			if 'duration' in time.attrib:
				key.duration = Time(time.attrib['duration'])
			self.times[key] = time.text.lower()
		if len(self.times) == 0:
			log('Error: no <time> tags found!')
			sys.exit()
		
		self.strings = {}
		stringsNode = root.find('strings')
		if stringsNode == None:
			log('Error: <strings> node not found!')
			sys.exit()
		for string in stringsNode.findall('string'):
			if 'id' not in string.attrib:
				log('Warning: found <string> tag with no "id" attribute')
				continue
			if string.text == None or string.text.strip() == '':
				log('Warning: empty <string> tag for id=%s' % string.attrib['id'])
				continue
			self.strings[int(string.attrib['id'])] = string.text.strip().lower()


class Theme:
	'''
	A theme has several elements that can be accessed through this class:
	* self.background  - The background color
	* self.active      - The color of enabled tiles
	* self.inactive    - The color of disabled tiles
	* self.image       - The filename of a background image, or None of omitted
	* self.imageWidth  - Image width, or 1280 if image is None
	* self.imageHeight - Image height, or 720 if image is None
	'''
	def __init__(self, file):
		log('Using theme: ' + os.path.basename(file))
		try:
			root = ElementTree.parse(file).getroot()
		except:
			log('Error parsing theme file!')
			raise # Let the user see the error
		try:
			self.active = root.find('active').text
			self.inactive = root.find('inactive').text
			imgNode = root.find('image')
			if imgNode != None:
				self.image = imgNode.text
				# Width and height default to fullscreen if unspecified
				self.imageWidth = int(imgNode.attrib['width']) if 'width' in imgNode.attrib else 1280
				self.imageHeight = int(imgNode.attrib['height']) if 'height' in imgNode.attrib else 720
				# If an image was found, background is optional
				if root.find('background') != None:
					self.background = root.find('background').text
				else:
					self.background = '00000000' # fully transparent
			else:
				# If an image wasn't found, background is mandatory
				self.background = root.find('background').text
				self.image = None
				self.imageWidth = 1280
				self.imageHeight = 720
			# In SS mode, the background is black so ignore the alpha channel
			if xbmc.getCondVisibility('System.ScreenSaverActive'):
				# Proceed if all chars are hex
				if self.isHexColor(self.background):
					self.background = 'ff' + self.background[2:]
		except:
			log('Error parsing theme file!')
			sys.exit()
	
	def isHexColor(self, color):
		isHex = lambda x: (ord('0') <= ord(x) and ord(x) <= ord('9')) or \
		                  (ord('A') <= ord(x) and ord(x) <= ord('F')) or \
		                  (ord('a') <= ord(x) and ord(x) <= ord('f'))
		return len(color) == 8 and [isHex(char) for char in color].count(False) == 0
